components/camel-solr/src/main/java/org/apache/camel/component/solr/SolrProducer.java

```
@@ -0,0 +1,302 @@
+import org.apache.solr.client.solrj.request.ContentStreamUpdateRequest;
+import org.apache.solr.client.solrj.request.DirectXmlRequest;
+import org.apache.solr.client.solrj.request.QueryRequest;
+import org.apache.solr.client.solrj.request.UpdateRequest;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.SolrInputDocument;
+import org.apache.solr.common.params.ModifiableSolrParams;
+
+/**
+ * The Solr producer.
+ */
+public class SolrProducer extends DefaultProducer {
+
+    public SolrProducer(SolrEndpoint endpoint) {
+        super(endpoint);
+    }
+
+    @Override
+    public void process(Exchange exchange) throws Exception {
+
+        String operation = (String) exchange.getIn().getHeader(SolrConstants.OPERATION);
+        if (operation == null) {
+            throw new IllegalArgumentException(SolrConstants.OPERATION + " header is missing");
+        }
+
+        // solr configuration
+        SolrConfiguration solrConfiguration = getEndpoint().getSolrConfiguration(operation);
+
+        // solr client
+        SolrClient solrClient = exchange.getIn().getHeader(SolrConstants.CLIENT, SolrClient.class);
+        if (solrClient == null) {
+            solrClient = getEndpoint().getComponent().getSolrClient(this, solrConfiguration);
+        }
+
+        // solr parameters
+        ModifiableSolrParams solrParams = new ModifiableSolrParams();
+        for (Map.Entry<String, Object> entry : exchange.getIn().getHeaders().entrySet()) {
+            if (entry.getKey().startsWith(SolrConstants.PARAM)) {
+                String paramName = entry.getKey().substring(SolrConstants.PARAM.length());
+                solrParams.add(paramName, entry.getValue().toString());
+            }
+        }
+
+        // solr collection
+        String solrCollection = exchange.getIn().getHeader(SolrConstants.COLLECTION, String.class);
```

I'm under the impression this is a mandatory header, no? Probably we should throw `NoSuchHeaderException` if this is missing as well.

---

org.apache.solr.client.solrj.impl.Http2SolrClient.request(SolrRequest<?>, String)
...
    if (ClientUtils.shouldApplyDefaultCollection(collection, solrRequest)) {
      collection = defaultCollection;
    }

There is defaultCollection in SolrClient. So its not mandatory in Exchange headers.